projects/igniteui-angular/src/lib/grids/grid/grid.component.ts

```diff
@@ -527,6 +535,204 @@ export class IgxGridComponent extends IgxGridBaseDirective implements GridType,
+     * ```
+	 * @memberof IgxGridComponent
+    */
+    public expandAll() {
+        const expandedStates = this.expansionStates;
+        this.data.forEach((rec) => {
+            expandedStates.set(this.primaryKey ? rec[this.primaryKey] : rec, true);
+        });
+        this.expansionStates = expandedStates;
+    }
+
+   /**
+     * Collapses all master rows.
+     * ```typescript
+     * this.grid.collapseAll();
+     * ```
+	 * @memberof IgxGridComponent
+    */
+    public collapseAll() {
+        this.expansionStates = new Map<any, boolean>();
+    }
+
+    /**
+     * Expands the master row by its id. ID is either the primaryKey value or the data record instance.
+     * ```typescript
+     * this.grid.expand(rowID);
+     * ```
+	 * @memberof IgxGridComponent
+     */
+    public expand(rowID: any) {
+        const expandedStates = this.expansionStates;
+        expandedStates.set(rowID, true);
+        this.expansionStates = expandedStates;
+    }
+
+    /**
+     * Collapses the master row by its id. ID is either the primaryKey value or the data record instance.
+     * ```typescript
+     * this.grid.collapse(rowID);
+     * ```
+	 * @memberof IgxGridComponent
+    */
+    public collapse(rowID: any) {
+        const expandedStates = this.expansionStates;
+        expandedStates.set(rowID, false);
+        this.expansionStates = expandedStates;
+    }
+
+
+    /**
+     * Toggles the master row by its id. ID is either the primaryKey value or the data record instance.
+     * ```typescript
+     * this.grid.toggle(rowID);
+     * ```
+	 * @memberof IgxGridComponent
+    */
+    public toggleRow(rowID: any) {
+        const expandedStates = this.expansionStates;
+        const state = expandedStates.get(rowID);
+        expandedStates.set(rowID, !state);
+        this.expansionStates = expandedStates;
+    }
+
+    public getDetailsContext(rowData, index) {
+        return {
+            $implicit: rowData,
+            index: index
+        };
+    }
+
+    public detailsKeyboardHandler(event, rowIndex, container) {
+        const colIndex = this.selectionService.activeElement ? this.selectionService.activeElement.column : 0;
+        const shift = event.shiftKey;
+        const ctrl = event.ctrlKey;
+        const key = event.key.toLowerCase();
+        const target = event.target;
+        if (key === 'tab') {
+            event.stopPropagation();
+            if (shift && target === container) {
+                // shift + tab from details to data row
+                event.preventDefault();
+                const lastColIndex = this.unpinnedColumns[this.unpinnedColumns.length - 1].visibleIndex;
+                this.navigateTo(rowIndex - 1, lastColIndex,
+                    (args) => args.target.nativeElement.focus());
+            } else if (!shift) {
+                const focusable = container
+                .querySelectorAll('button, [href], input, select, textarea, [tabindex]:not([tabindex="-1"])');
```

Comment:
@radomirchev @rkaraivanov 

Alternative to this code we discussed the idea of having a hidden element at the end to take the focus.

However, after discussing this with @MayaKirova , I think a simpler solution with not much worse end-user experience is to focus the outer details container regardless of where the focus is coming from. 

This will look like this:

- Last cell <kbd>Tab</kbd> Details Container  <kbd>Tab</kbd> Details Container contents ....  <kbd>Tab</kbd> First cell
- First cell <kbd>Shift</kbd> + <kbd>Tab</kbd> Details Container <kbd>Shift</kbd> + <kbd>Tab</kbd> Last cell
- First cell <kbd>Shift</kbd> + <kbd>Tab</kbd> Details Container <kbd>Tab</kbd> Details Container contents... 

Basically <kbd>Shift</kbd> + <kbd>Tab</kbd> won't drill down through the details contents, the users will drill down with <kbd>Tab</kbd> in it (or <kbd>Shift</kbd> + <kbd>Tab</kbd> but only while inside)